metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -784,15 +815,32 @@ public void handleLeaderChange(LeaderAndEpoch newLeader) {
                             curEpoch);
                     }
                     log.info(
-                        "Becoming the active controller at epoch {}, committed offset {} and committed epoch {}.",
-                        newEpoch, lastCommittedOffset, lastCommittedEpoch
+                        "Becoming the active controller at epoch {}, committed offset {}, committed epoch {}, and metadata.version {}",
+                        newEpoch, lastCommittedOffset, lastCommittedEpoch, activeMetadataVersion
                     );
 
                     curClaimEpoch = newEpoch;
                     controllerMetrics.setActive(true);
                     writeOffset = lastCommittedOffset;
                     clusterControl.activate();
 
+                    // Check if we need to bootstrap a metadata.version into the log. This must happen before we can
+                    // write any records to the log since we need the metadata.version to determine the correct
+                    // record version
+                    if (activeMetadataVersion == MetadataVersions.UNINITIALIZED.version()) {
+                        if (initialMetadataVersion == MetadataVersions.UNINITIALIZED) {
+                            prependWriteEvent("initializeMetadataVersion", () -> {
+                                log.info("Upgrading from KRaft preview. Initializing metadata.version to 1");
+                                return featureControl.initializeMetadataVersion(MetadataVersions.V1.version());
```

Comment:
Always initialize to version 1 when upgrading from existing (preview) KRaft cluster